src/util/address.rs

```diff
@@ -71,30 +79,40 @@ pub enum Error {
     InvalidSegwitV0ProgramLength(usize),
     /// An uncompressed pubkey was used where it is not allowed.
     UncompressedPubkey,
+    /// Witness version which is not (yet) supported
+    UnsupportedWitnessVersion {
+        /// Witness version that was found during address decoding
+        version: WitnessVersion,
+        /// Length of the witness program
+        length: usize
+    },
     /// Address size more than 520 bytes is not allowed.
     ExcessiveScriptSize
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match *self {
-            Error::Base58(ref e) => write!(f, "base58: {}", e),
-            Error::Bech32(ref e) => write!(f, "bech32: {}", e),
-            Error::EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
+            Error::Base58(_) => f.write_str("base58 address encoding error"),
+            Error::Bech32(_) => f.write_str( "bech32 address encoding error"),
+            Error::EmptyBech32Payload => f.write_str( "the bech32 payload was empty"),
+            Error::InvalidBech32Variant { expected, found } => write!(f, "invalid bech32 checksum variant found {:?} when {:?} was expected", found, expected),
             Error::InvalidWitnessVersion(v) => write!(f, "invalid witness script version: {}", v),
-            Error::UnparsableWitnessVersion(ref e) => write!(f, "Incorrect format of a witness version byte: {}", e),
+            Error::UnparsableWitnessVersion(_) => f.write_str( "incorrect format of a witness version byte"),
             Error::MalformedWitnessVersion => f.write_str("bitcoin script opcode does not match any known witness version, the script is malformed"),
-            Error::InvalidWitnessProgramLength(l) => write!(f,
-                "the witness program must be between 2 and 40 bytes in length: length={}", l,
+            Error::InvalidWitnessProgramLength(len) => write!(f,
+                "the witness program must be between 2 and 40 bytes in length: length={}", len,
```

Comment:
One more general thought on error design, not related to this PR specifically: something like [`serde::de::Visitor::expecting`](https://docs.serde.rs/serde/de/trait.Visitor.html#required-methods) method or [`parse_arg::ParseArg::describe_type`](https://docs.rs/parse_arg/0.1.4/parse_arg/trait.ParseArg.html#required-methods) could be useful.

Reply:
Not sure I entirely get what do you mean by that design

Reply:
Separate what is expected from what is the error. The intention is mainly to allow nicer/more flexible formatting of error messages.